dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
+length of the text before the change is CHANGE-BEFORE-LENGTH."
+  (combine-after-change-calls
+    (dart-analysis-server-send
+     "analysis.updateContent"
+     `((files
+        . ((,buffer-file-name
+            . ((type . "change")
+               (edits
+                . (((offset . ,change-begin)
+                    (length . ,change-before-length)
+                    (replacement
+                     . ,(buffer-substring change-begin change-end)))))))))))))
+
+(defun dart-remove-analysis-overlay ()
+  "Remove the overlay for the current buffer since it has been saved."
+  (dart-analysis-server-send
+   "analysis.updateContent"
+   `((files . ((,buffer-file-name . ((type . "remove"))))))))
+
+(defun dart-add-current-file-to-analysis-roots ()
+  "Add the current file's pub root or directory to analysis roots."
+  (let ((pub-root (locate-dominating-file buffer-file-name "pubspec.yaml"))
+        (current-dir (file-name-directory buffer-file-name)))
+    (if pub-root (dart-add-to-analysis-roots (expand-file-name pub-root))
+      (dart-add-to-analysis-roots (expand-file-name current-dir)))))
+
+(defun dart-add-to-analysis-roots (dir)
+  "Add DIR to the analysis roots."
+  (add-to-list 'dart-analysis-roots dir)
+  (dart-set-analysis-roots))
+
+(defun dart-set-analysis-roots ()
+  "Send the current list of analysis roots to the analysis server."
+  (dart-analysis-server-send
+   "analysis.setAnalysisRoots"
+   `(("included" . ,dart-analysis-roots)
+     ("excluded" . nil))))
+
+(defun dart-analysis-server-send (method &optional params callback)
+  "Send the METHOD request to the server with optional PARAMS.
+
+If you provide a CALLBACK, it will be called with the JSON decoded
+response.  Otherwise, the output will just be checked."
+  (let ((req-without-id (dart-analysis-server-make-request method params)))
+    (dart-analysis-server-enqueue req-without-id callback)))
+
+(defun dart-analysis-server-make-request (method &optional params)
+  "Construct a request for the analysis server.
+
+The constructed request will call METHOD with optional PARAMS."
+  `((method . ,method) (params . ,params)))
+
+(defun dart-analysis-server-enqueue (req-without-id callback)
+  "Send REQ-WITHOUT-ID to the analysis server, call CALLBACK with the result."
+  (setq dart-analysis-server-next-id (1+ dart-analysis-server-next-id))
+  (let ((request
+         (json-encode (push `(id . ,(format "%s" dart-analysis-server-next-id))
+                            req-without-id))))
+    (dart-log (concat "Sent:\n" request))
+    (if callback
+        (push `(,dart-analysis-server-next-id . ,callback)
+              dart-analysis-server-callbacks)
+      (push
+       `(,dart-analysis-server-next-id . dart-analysis-server-no-error-callback)
+       dart-analysis-server-callbacks))
+    (process-send-string (dart-analysis-server-process dart-analysis-server)
+                         (concat request "\n"))))
+
+(defun dart-analysis-server-no-error-callback (response)
```

Comment:
It looks like this is supposed to be `on-error-callback` rather than `no-error-callback`.